### ISSUE TITLE
Raspberry Pi OS Bullseye not supported due to lack of Python2 MySQL libraries

### ISSUE THREAD

**Issue by mexon** [ansible-collections/ansible.mysql#272]
##### SUMMARY
I've been trying to run my playbook that contains this task:

```
- name: update mysql root password for all root accounts
  mysql_user: 
    name: root 
    host: "{{ item }}" 
    login_user: root
    check_implicit_admin: yes
    priv: "*.*:ALL,GRANT"
```

It produces this error message:

```
failed: [test1] (item=test1) => {"ansible_loop_var": "item", "changed": false, "item": "test1", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
failed: [test1] (item=127.0.0.1) => {"ansible_loop_var": "item", "changed": false, "item": "127.0.0.1", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
failed: [test1] (item=::1) => {"ansible_loop_var": "item", "changed": false, "item": "::1", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
failed: [test1] (item=localhost) => {"ansible_loop_var": "item", "changed": false, "item": "localhost", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
```

I tried to fix it by changing the earlier package installation step by changing `python-mysqldb` to `python3-mysqldb`, and when that didn't work by adding `python3-pymysql`:

```
- name: install MySQL server
  apt:
    name:
    - mariadb-server
    - python3-mysqldb
    - python3-pymysql
    update_cache: yes
    cache_valid_time: 259200
  become: yes
```

I also tried updating the version of community.mysql from 2.3.1 to 3.0.0.  But it still wasn't working.

It seems that community.mysql requires Python2 to be present.  Eventually I solved the problem by retrieving a `.deb` package from an older version of Raspberry Pi OS and installing that.  This is not a good solution though.  Python2 is end of life, that's why the python2 versions of these libraries have been removed from Raspberry Pi OS.

As a side issue, the error message is ambiguously worded.  One interpretation: I need PyMySQL if I use Python3, or either PyMySQL or MySQL-python if I use Python2.  Another interpretation: I need to either install both Python2 and Python3 versions of PyMySQL, or I need to install the Python2 version of MySQL-python.  Please clarify the wording.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
`mysql_user`

##### ANSIBLE VERSION
```
ansible-playbook [core 2.11.6] 
  config file = None
  configured module search path = ['/Users/mat/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/4.8.0/libexec/lib/python3.10/site-packages/ansible
  ansible collection location = /Users/mat/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.10.0 (default, Oct 13 2021, 06:45:00) [Clang 13.0.0 (clang-1300.0.29.3)]
  jinja version = 3.0.2
  libyaml = True
```

##### COLLECTION VERSION
```# /usr/local/Cellar/ansible/4.8.0/libexec/lib/python3.10/site-packages/ansible_collections
Collection      Version
--------------- -------
community.mysql 2.3.1  

# /Users/mat/.ansible/collections/ansible_collections
Collection      Version
--------------- -------
community.mysql 3.0.0  
```

##### CONFIGURATION
```
[EMPTY]
```

##### OS / ENVIRONMENT
macOS Big Sur 11.6.2
Homebrew 3.3.9

##### STEPS TO REPRODUCE
See main description

##### EXPECTED RESULTS
```
TASK [update mysql root password for all root accounts] ******************************************************************************************************
changed: [test1] => (item=test1)
changed: [test1] => (item=127.0.0.1)
changed: [test1] => (item=::1)
ok: [test1] => (item=localhost)
```

##### ACTUAL RESULTS
```
TASK [update mysql root password for all root accounts] ******************************************************************************************************
failed: [test1] (item=test1) => {"ansible_loop_var": "item", "changed": false, "item": "test1", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
failed: [test1] (item=127.0.0.1) => {"ansible_loop_var": "item", "changed": false, "item": "127.0.0.1", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
failed: [test1] (item=::1) => {"ansible_loop_var": "item", "changed": false, "item": "::1", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
failed: [test1] (item=localhost) => {"ansible_loop_var": "item", "changed": false, "item": "localhost", "msg": "The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required."}
```

**Comment by Andersson007** [comment]
@mexon hello, thanks for reporting this!
When working with Python 3 libs, did you try to change the `ansible_python_interpreter` variable https://docs.ansible.com/ansible/latest/reference_appendices/python_3_support.html#:~:text=Ansible%20will%20automatically%20detect%20and,%2Fusr%2Fbin%2Fpython3. ? If not, could you please try and tell us if it helps.

**Comment by mexon** [comment]
I tried adding `test1 ansible_python_interpreter=/usr/bin/python3` in my inventory, and indeed it no longer complains about a missing package.  Now it fails with `"unable to find /root/.my.cnf. Exception message: (1698, \"Access denied for user 'root'@'localhost'\")"`.  I don't have time to investigate that further, but it seems like setting the variable is a good idea.  Can the error message be changed to suggest doing that?

**Comment by Andersson007** [comment]
@mexon thank you for the feedback!
1. Which module produced this error message? One of mysql_db.py, mysql_variables.py, mysql_replication?
2. Has passing the path to the file helped?

Thanks!

**Comment by mexon** [comment]
It's mysql_db.py.  I tried `ANSIBLE_KEEP_REMOTE_FILES=1` and I have the temporary file called `AnsiballZ_mysql_db.py`.  It's got passwords and stuff in so probably not a good idea to post it here.

From a quick look at it, I would guess it's related to this new thing where by default, mariadb does not allow login as root.  Instead you have to sudo and login without any credentials.  That is:

```
admin@test1:~ $ mysql -u root -p
Enter password: 
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
admin@test1:~ $ mysql -u root -h localhost -p
Enter password: 
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
admin@test1:~ $ sudo mysql -u root
Welcome to the MariaDB monitor.  Commands end with ; or \g.
```

The playbook I'm using this time is different, since the original task I reported succeeds:

```
---
- hosts: friendica
  remote_user: admin
  become: yes
  tasks:
  - name: install MySQL server
    apt:
      name:
      - mariadb-server
      - python3-mysqldb
      update_cache: yes
      cache_valid_time: 259200
  - name: Create friendica database
    mysql_db:
      name: friendica
      state: present
      login_user: root
    become: yes
```

Possibly I'm not supposed to change the root account as in the original playbook?  I just noticed that I didn't show the whole thing above. I probably copy pasted this from the internet somewhere.  Anyway, it used to work, and still does with python 2.

```
- name: update mysql root password for all root accounts
  mysql_user: 
    name: root 
    host: "{{ item }}" 
    login_user: root
    check_implicit_admin: yes
    priv: "*.*:ALL,GRANT"
  with_items:
    - "{{ ansible_hostname }}"
    - 127.0.0.1
    - ::1
    - localhost
  become: yes
```

**Comment by Andersson007** [comment]
@mexon thanks for the feedback!
1. I think we should definitely mention the possible issues (same as you get) and  elaborate a bit more on the solution in the NOTES section of https://github.com/ansible-collections/community.mysql/blob/main/plugins/doc_fragments/mysql.py. Yes, there's some related info already there but it's not enough, IMO. Would you like to create a PR? I would add it to the second paragraph. This is a frequently reported issue, so the PR would be helpful. What do you think?

2. I feel extremely dumb today, sorry if one of the following question is stupid - do you have any issues now?:) If not, is there a way we can improve doc? If yes, i would suggest closing this issue after we update the doc as I'm suggesting above and reopen a new one.

Thank you!

**Comment by mexon** [comment]
Well, now that I actually _read_ the notes, I see that they do suggest changing the interpreter, and provide a link to the documentation 😳  So there's nothing much to add there, but I followed your suggestion and submitted a PR that expands the notes slightly.  I also expanded the error message, which I'd argue is the more important place to put hints.  I'm not the only person out there who submits the bug report before reading the notes...

Those notes also give a hint as to what's happening after I switch to Python3.  If there are no credentials, it tries `~/my.cnf`, and if that fails it should fall back to just using no password.  That last one should work, but instead it seems to give up when the `~/my.cnf` thing fails.  That feels like it might be worth more investigation.  Separate issue I guess.

I don't really have issues any more, in that I have the workaround of installing the old python 2 library.  As such, I would like to return to fixing the actual bug in friendica I was trying to fix when I got distracted by this problem 😉  But if you think I should pursue the authentication problem when using python 3, I can do that.
